### DISCUSSION TITLE
로컬 환경에서 H2 DB 사용

### DISCUSSION THREAD

**Discussion by Choi-JJunho** [BCSDLab/KOIN_API_V2#150] [에러 / 고민사항]
> 디스코드 채널에서 나온 질문을 옮긴 discussion입니다.

# 안건
로컬환경에선 테스트 컨테이너 대신에 H2 이용하는 건에 대해

## 사유

로컬에서 테스트 수행 시 docker 구동으로 인해 테스트 속도가 느리다.
깃헙 테스트에서만 테스트 컨테이너 쓰는 방법은 어떤가?

참고자료: https://www.youtube.com/watch?v=fPJ7EJM1a0k

## 정리

로컬에서 테스트를 구동할때는 H2 DB를 사용하고 Github Action CI 환경에서는 TestContainer로 구동하는 환경을 구성하는 방안에 대한 고민

cc. @Invidam @songsunkook @daheeParkk

**Comment by Choi-JJunho** [comment]
로컬 H2 설정이 추가되면 관리포인트가 늘어나게 될 것 같네요
개인적으로 관리가 편한것이 협업 과정에 더 유리하다고 생각합니다.

신규 부원이 6개월마다 들어오는 동아리 특성 상 관리 및 설정이 편리한 것을 선호합니다.

다른 인원들의 로컬은 잘 모르겠지만 현재 제 로컬에서 테스트는 약 4초정도 소요됩니다.
<img width="506" alt="image" src="https://github.com/BCSDLab/KOIN_API_V2/assets/49794401/31321d5f-b341-43ee-9dec-69776ce4e23c">

인수테스트임을 감안했더라도 4초 정도의 속도면 개발에 지장은 없을 것으로 여겨지네요

따라서 `현재 상황에서는 로컬에서 TestContainer를 돌려도 무리가 없어 보인다.` 라는 결론입니다.

**Reply by songsunkook** [reply]
저는 로컬 환경 기준으로 도커에 연결되는 시간이 약 8~10초 걸리는 것으로 보입니다.

```
11:31:42.294 [main] INFO tc.mysql:5.7.34 -- Waiting for database connection to become available at jdbc:mysql://localhost:62248/test using query 'SELECT 1'
11:31:50.757 [main] INFO tc.mysql:5.7.34 -- Container mysql:5.7.34 started in PT8.603926S
```
(`SELECT 1` 쿼리에 대한 응답까지의 시간 약 8초 경과)

도커 연결 이후 테스트 시간은 말씀하신 것처럼 2 ~ 5초 내로 완료되는 것을 확인했는데, 실제 테스트 시간 외에 도커 연결 시간이 부담스러울 수 있을 것 같습니다. 이 시간은 테스트 수와 무관하게 항상 8 ~ 10초 가량 걸리기 때문에 단일 테스트 실행 시에도 매 테스트마다 10초가량이 소모됩니다.(윈도우의 경우 더 소모될 것으로 예상합니다.)

따라서 저는 `로컬 H2 설정을 추가하면 테스트에 소모되는 시간은 유의미하게 줄어들 것`이라고 생각합니다.

+) 제가 로컬 환경 설정을 잘못한 부분이 있거나 로그를 잘못 해석한 부분이 있다면 말씀해주시면 감사하겠습니다.

**Reply by Choi-JJunho** [reply]
docker container가 구동되는 환경을 미처 생각하지 못했네요!
말씀하신대로 테스트를 구동할 시 10초 이상의 시간이 소요되는 것으로 파악이되는군요.
속도 측면에서는 확실히 이점을 얻을 수 있을것이라는 생각이 드네요.

다만 다음 상황에 대해 추가적인 고민이 들어 아래 질문에 답변해주시면 감사하겠습니다.

1. Redis는 어떻게 테스트할건가?
2. H2에서 mode=MySQL 옵션이 있긴하지만 MySQL 5.7버전과 쿼리 호환이 100% 일치하는가?
3. 인수테스트를 수행하는 목적이 무엇인가?

**Reply by Invidam** [reply]
비교하기위해서 어디에서 시간이 많이 소요되는지 파악이 필요할 것 같아요 

기억상으로는 테스트 컨테이너 올리는 시간, mysql 컨테이너 올리는 시간, redis 컨테이너 올리는 시간, DB 초기화하는 시간, 테스트하는 시간

중에 어떤 것 때문에 시간 소요가 많이 되는지..

**Reply by Invidam** [reply]
추가적으로 mysql 5.7이라서 (애플 실리콘이 아니라서) 느린건가?? 의문도 들더라구요 mysql 8.0 버전대는 얼마나 걸리는 지도 테스트 함 해주시면 좋을 것 같아요.
